[Wallet] Enable RBF by default in QT

Sjors opened the pull request:
~If there are no objections, this would supersede #11556.~

Enabling RBF by default avoids the need to explain all possible use cases of RBF.

This PR does not change the default RPC wallet behavior, as this could break implementations that depend on it and it's not clear what happens when automated services suddenly switch on RBF on a large scale.

After trying various approaches, we settled on just having QT ignore `-walletrbf`.

Send screen:
<img width="388" alt="send" src="https://user-images.githubusercontent.com/10217/34251097-329c8dee-e63f-11e7-9e14-d7f55d2b52cc.png">


Confirmation screen by default (with RBF):
<img width="429" alt="rbf yes" src="https://user-images.githubusercontent.com/10217/32442799-f50d54aa-c2fc-11e7-9392-96339d0f1f74.png">

Confirmation screen without RBF:
<img width="431" alt="rf no" src="https://user-images.githubusercontent.com/10217/32442793-ef30bc34-c2fc-11e7-8ca2-e86a97175278.png">

Sjors commented:
Requesting Travis restart. Most of the environments passed, three failed for `txn_clone.py`, which seems unrelated.

TheBlueMatt commented:
Concept ACK. I believe the test failures are, in fact, because of the change in default.

Sjors commented:
@TheBlueMatt ok, I'll double check that. Any theory why they would not fail on _all_ machines (including my own)?

TheBlueMatt commented:
listtransactions fails for me locally, too.

MeshCollider commented:
@Sjors not all the travis machines run the full python test suite, each one serves a different purpose, but those three failing definitely means its related to this change :)

MeshCollider commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -342,12 +342,17 @@ void SendCoinsDialog::on_sendButton_clicked()
+      questionString.append(tr("You can increase the fee later (signals Replace-By-Fee)."));
     }
+    else
+    {
```
Nit, braces on same line as `if` and `else`

MeshCollider commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -342,12 +342,17 @@ void SendCoinsDialog::on_sendButton_clicked()
-        questionString.append("<hr /><span>");
-        questionString.append(tr("This transaction signals replaceability (optin-RBF)."));
-        questionString.append("</span>");
+      questionString.append(tr("You can increase the fee later (signals Replace-By-Fee)."));
```
Nit, indentation
Also perhaps add reference to BIP 125 again

Sjors replied:
Will do.

meshcollider reviewed (commented):
Number of unrelated whitespace changes too, you can disable your editor from doing that by default :)

MarcoFalke commented:
Please don't tag with [qt], when you are actually modifying the default for all wallet created transactions

MarcoFalke commented:
Would also be helpful to have some reasoning in OP that outlines how this will not cause user dissatisfaction because some merchants handle rbf txes differently.

Sjors commented:
@MarcoFalke I'll try if I can refactor this in a way that in only impacts QT. Changing the default behavior of RPC is probably not a good idea, since there are automated integrations.

The current implementation also doesn't warn users that some merchants handle RBF differently, but this may indeed be useful when it becomes the default. Different is not always worse by the way; e.g. ShapeShift will show you a recommended fee to bump to if you want a tx to confirm faster.

@MeshCollider good to know not all Travis machines run the same tests. Does `make test` not cover everything either?

Sjors commented:
@MarcoFalke I'll take a look at how some merchants and payment processors currently handle RBF, and add that to my PR description. Unless someone has already done that recently and can link to the research...

promag commented on src/wallet/wallet.h:
```diff
@@ -262,7 +262,7 @@ class CMerkleTx
     bool IsCoinBase() const { return tx->IsCoinBase(); }
 };
 
-/** 
```
Remove this whitespace changes.

promag reviewed (commented):
This PR includes 2 different things: change the default RBF value and change the UI. I was expecting to not see the UI change. I think it is enough to change `DEFAULT_WALLET_RBF`, update the current checkbox (should be checked) and fix the tests. The UI change could still be done in #11556.

Edit: agree that [qt] prefix doesn't apply.

Sjors commented:
Alternatively I could leave `DEFAULT_WALLET_RBF` alone (not touching RPC behavior), add a new DEFAULT_WALLET_UI_RBF and toggle that. In that case the QT label would be correct.

The changes in  #11556 require very different wording imo. If you disable a feature by default you need to persuade users why they might want to opt-in to this functionality. You're selling them on the feature. If you enable it by default, then you need to explain why someone might object to it. You might even move the check box somewhere else in the UI, under some "advanced" section.

MeshCollider commented:
@Sjors no, `make test` doesn't run the functional python tests I don't think, you have to run test_runner.py for that :)

jonasschnelli commented:
General Concept ACK. Most payment providers do wait for confirmations anyways so I think it makes sense now to use BIP125-RBF-signaling as a default. Conforming to an eventually (and unsafe) 0-conf payment receiver, one can then set the RBF disable flag.

Code wise, this is incomplete.
I think introducing a UI based default makes little sense. IMO we should continue this with...
* fix `sendtoaddress`, `sendmany`, `fundrawtransaction`, `bumpfee` and swap the default there (rename from enable to disable)
* It's an API change, so please add a release notes part

MarcoFalke commented:
I am fine if you change the default globally. But then remove the qt tag
and add some release notes.

promag commented:
> (rename from enable to disable)

@jonasschnelli what should be renamed (the RPC option is `replaceable`)?

> It's an API change, so please add a release notes part

Is it an API change because the default value changed? The default value can already be changed with `-walletrbf`. This also happens withm for instance, `-keypool` and `keypoolrefill` RPC. Agree in adding release note but more regarding `-walletrbf` and the impact in coin control, these RPC and the wallet transactions.

I still think the UI change is independent of this, and subject to other discussion (aim for easier reviews).

jonasschnelli commented:
Things like....
```
\"replaceable\"       (boolean, optional, default true) Whether the new transaction should still be\n"
            "                         marked bip-125 replaceable. If true, the sequence numbers in the transaction will\n"
            "                         be left unchanged from the original. If false, any input sequence numbers in the\n"
            "                         original transaction that were less than 0xfffffffe will be increased to 0xfffffffe\n"
            "                         so the new transaction will not be explicitly bip-125 replaceable (though it may\n"
            "                         still be replaceable in practice, for example if it has unconfirmed ancestors which\n"
            "                         are replaceable).\n"
```

This change does also change the default of the RPC layer which may have wide consequences for existing users.
If it's an API change or a local policy change is questionable. However, upgrading can have wide consequences for wallet users.

I think we should always mention to default state (based on `-walletrbf`?) in the RPC help commands.

promag commented:
Yes, that is missing.

Sjors commented:
I'm not comfortable (yet) changing code at the RPC level. So what I can do is change this PR to work only at the QT level. Someone else can make a PR that changes it all the way down, which would supersede this. Then you can all choose between #11556, this and that other PR.

Sjors commented:
Update the PR: there's now a new setting `DEFAULT_WALLET_QT_RBF` which is `true`. This setting is used by BitcoinQT, unless it's launched with `-walletrbf`. RPC behavior remains unchanged.

Whitespace is gone now, but I'm still going through the other feedback.

Sjors commented:
Rebased. Tests work. Updated screenshots. Explicitly mention BIP-125.

ryanofsky commented:
Haven't really looked at this yet, but there was a previous PR to try to switch the rbf default (for both qt and rpc) in #9527, and some of that discussion might apply here.

Sjors commented:
@ryanofsky that previous PR turned on RBF by default for both the wallet and RPC.

A number of objections were raised. I'll address the ones that I think are relevant:

@luke-jr (and @instagibbs):
> Probably shouldn't be merged until something can actually use it (preferably from the GUI).

That's the case now.

@RHavar:
> I was supportive of bip125 because it was marketed, sold and even titled: "Opt-in". But this simple change in defaults really makes it "Opt-out" and will end up causing huge impacts to people who are currently relying on 0-conf transactions, as suddenly everyone is going to be accidentally sending "opt-in" RBF.

I suppose that's unchanged.

There were some issues with usability at the time, which I'm guessing have been addressed since?

There was also discussion about RBF vs. CPFP. Given that the RBF check box made it into the UI, I don't think that's relevant (at least not for the scope of this issue). Although perhaps we could warn users that using RBF breaks the ability of CPFP for the recipient (does it?). I could imagine a merchant using CPFP for large purchases as a customer service gesture.

@TheBlueMatt:
> Obviously I'd agree we shouldn't do this unless bumpfee goes in.

That's there now. 

@gmaxwell:
> As far as your zero-conf comments, that is why my comment here and on the bumpfee PR talk about being able to after-the-fact turn it off.

Is this currently possible? If not, should we wait for that?

@RHavar:
> If this is going to be the default, it seems that bumpfee first at least be able to handle all transactions (e.g. ones without change)?

I don't understand RBF really well, but if transactions without change can't use RBF, it would be good UX to grey-out and the checkbox for those transactions (and set it to disabled). We don't want the wallet to throw errors with default settings. Does it mean you can't send the transaction at all, or just not bump it? Both are problematic though.

> Likewise it would also make sense to have "reverttransaction" as a prerequisite too. It would help people fully realize how revertible bip125 transactions are

I'd like that feature too, but I think it can be added later.

> promoting the controversial feature as "opt-in" (which is even in the name of the BIP) and now quietly being changing it to "opt-out".

I think opt-in refers to the fact that it's a soft fork? I think the best way to counter bad press is to make the feature work great and have users and (most) merchants love it.

By the way, the bump fee feature is pretty well hidden in the UI, but that's another issue...

So some original objections might still stand, but others have been addresses since early this year. So it seems useful to discuss this again.

MarcoFalke commented on src/wallet/wallet.h:
```diff
@@ -63,6 +64,7 @@ static const bool DEFAULT_WALLET_REJECT_LONG_CHAINS = false;
 static const unsigned int DEFAULT_TX_CONFIRM_TARGET = 6;
 //! -walletrbf default
 static const bool DEFAULT_WALLET_RBF = false;
+static const bool DEFAULT_WALLET_QT_RBF = true;
```
Eh. I guess we don't want Qt code to end up here

Sjors replied:
I'm not sure how to disentangle this. The `-walletrbf` is cast to a boolean in init.cpp, so QT can't tell if that flag was present or not.

laanwj replied:
What about a separate option for the GUI, GUI-only, that defaults to `-walletrbf`? This would allow it to be completely disentangled from the core.

MarcoFalke commented on src/qt/walletmodel.cpp:
```diff
@@ -754,5 +754,5 @@ int WalletModel::getDefaultConfirmTarget() const
 
 bool WalletModel::getDefaultWalletRbf() const
```
No need for this wrapper if the member is qt-only

MarcoFalke commented:
utACK 58250a5567e8d99cd00321a2c33b30d141b2935d

RHavar commented:
> I don't understand RBF really well, but if transactions without change can't use RBF, it would be good UX to grey-out and the checkbox for those transactions (and set it to disabled). 

That's pretty fragile and complex. You need to run coin-selection to know if the transaction is going to have no change or not. I strongly suspect it'll be easier to just fix the cases of bumping transactions without change

MarcoFalke commented:
> fix the cases of bumping transactions without change

Should be a separate pull, though.

instagibbs commented:
@RHavar I think he's referring to the "bump fee" button itself, not the "enable bip125" button?

RHavar commented:
@instagibbs he did say "checkbox" though, which makes it sound like it's the original transaction (and from a users point of view it doesn't really make sense to use bip125 for a transaction that you can't fee bump anyway). But a setting "enable bip125 (if we have change)" is the logical behavior, but annoying to explain.

--

All that said, I'm still very against this change. The reiterate my objections:

* bip125 was explicitly sold as "opt-in" to turn it to "opt-out" is rather distasteful, and has a lot of externalities. 

I would drop this objection if there was a significant (>5%) of hash power that was maliciously mining 0-conf double spends. But there simply isn't.

My latest project has in the last week processed 279 transactions worth 339 BTC with 0-confirmations transaction (after passing a few safety checks) without a single successful double-spend. I'm aware it's probably going to break horribly in the future, but it would be appreciate if yall didn't go out of your way to break shit that works and gives users a good experience.

And my other objection is, that core users can't even take full advantage of bip125. Unless there's an exposed simple "bump fee" (that works reliably) and "revert transaction". It seems totally premature to expose users (and the rest of the bitcoin ecosystem) to all the disadvantage of bip125, without at the very least allowing users fully take advantage of it.

Sjors commented:
@RHavar I'm still trying to understand why RBF would break your application. If you trust customers to not double spend coins, why would they behave any different with RBF enabled? Just because it's marginally easier to cheat?

RHavar commented:
> If you trust customers to not double spend coins

I don't trust them. I don't have any of their details, and if they could successfully double spend even 2% of the time, they could turn a profit. They would require the assistance of a malicious miner (who could do it costlessly). However no miners are doing it at the moment (and I hope this stays for a long time; even though it probably won't).

Sjors commented:
So your concern is that RBF makes it easier to double spend? But the feature is already there. I would expect a customer with malicious intent to just tick the box. Or are you worried about a more opportunistic thief; someone who didn't realize this was possible, but because they see the bump-fee feature, suddenly they start reading up on BIP-125, spin up the RPC client and redirect their transaction?

Sjors commented:
Users can't use bump fee if they didn't tick the box. I suspect they won't even know that they could have used RBF until they experience a stuck transaction for the first time. The wallet recommends lower fees when RBF is enabled, but again, users don't know that and so they pay higher fees despite the RBF feature being available. That drives the fee up for everyone (not if QT is only a small fraction of users, but larger wallets might be reluctant to support RBF if even Core doesn't have it on by default).

What is the issue with the bump fee implementation you're referring to?

It seems the trade-off here is such that users either get a bad experience because their transaction is stuck for ages, or a bad experience because their merchant has to wait for confirmations. I don't know which one is more common, so I don't know what the average impact is. In terms of worst case, having to wait a few blocks is far better than having to wait several days. So this may be a situation of slightly worse average case, but significantly less bad worst case.

CPFP seems of out of scope, but it's also quite expensive; unless you needed to move the received funds anyway, you end up paying for two transactions rather than just one.

RHavar commented:
> Users can't use bump fee if they didn't tick the box. I suspect they won't even know that they could have used RBF until they experience a stuck transaction for the first time. 

I think this is more of an argument for supporting a sort of limited RBF in all cases as long as all output amounts are unchanged or increased.

>  users don't know that and so they pay higher fees 

I think this is more of a UI issue. When showing a fee it could give a prediction for different percentiles of probability and explain the transaction is final or not.

> What is the issue with the bump fee implementation you're referring to?

You can't bumpfee on a transaction that doesn't have change (this is just a pure implementation limitation) or a transaction in which you or the receiver has spent  (which makes sense, there's a lot complexity here) 

> CPFP seems of out of scope

Why? It's often a good alternative to what you are suggesting. Especially if you don't want to (or can't) break a transaction chain (e.g. the receiver has already spent your funds, but it hasn't confirmed)

>  but it's also quite expensive; unless you needed to move the received funds anyway

The sender can do it too if there's change. Core's coinselection is sufficiently unoptimized that forcing the use of particular inputs has almost always 0 cost. But yeah, you generally need to be wanting to make a new transaction anyway for it to be cost effective.

petertodd commented:
Re: @RHavar 

> "I don't have any of their details, and if they could successfully double spend even 2% of the time, they could turn a profit. However, they would require the assistance of a malicious miner (who could do it costlessly). However no miners are doing it at the moment (and I hope this stays for a long time; even though it probably won't)."

Not sure what exactly your service actually is, but not only do more than 2% of miners run full-RBF - and thus will replace a transaction even if it hasn't set the opt-in RBF bit - a significant chunk of the ones who don't have sufficiently non-standard "IsStandard()" rules that they might as well be running full-RBF.

petertodd commented:
Concept ACK

RHavar commented:
> Not sure what exactly your service actually is, but not only do more than 2% of miners run full-RBF - and thus will replace a transaction even if it hasn't set the opt-in RBF bit

I haven't observed this at all. What miners or mining pools are doing this? Are you sure you're not confusing a mining pool that uses a reasonably small mempool with full-RBF?

I did see some full-RBF behavior a while back, I think it was a year or so (?) ago (going by memory) but it was very short-lived (like for 2 or 3 days max)

> a significant chunk of the ones who don't have sufficiently non-standard "IsStandard()" rules that they might as well be running full-RBF

Yeah, I've noticed this. But it's something that that I haven't had too much trouble accounting for

petertodd commented:
@RHavar Are you running a full-RBF node? E.g. https://github.com/petertodd/bitcoin/tree/replace-by-fee-v0.15.0.1 Quite possible that you're just not noticing it.

I say they're running full-RBF, not because I've personally witnessed it, but rather based on the pools that have reached out to me and told me they're running it (including running my full-rbf fork).

> Yeah, I've noticed this. But it's something that that I haven't had too much trouble accounting for

How do you account for it?

instagibbs commented:
>It seems the trade-off here is such that users either get a bad experience because their transaction is stuck for ages, or a bad experience because their merchant has to wait for confirmations.

Agreed. We should just document this well so merchants can tell their users why their deposits are delayed, and how to uncheck it if they want.

concept ACK

RHavar commented:
> Quite possible that you're just not noticing it.

I quite often see RBF attempts, just never see them working

> I say they're running full-RBF, not because I've personally witnessed it

Well I'm actually actively monitoring for it, and not seeing it. If it's happening, it should be immediately obvious and scammers will be quick to take advantage of it. In fact @petertodd you have my full permission to take advantage of up to 0.3 BTC by abusing 0-confs on www.dustdice.com . You can consider it a bounty =)  All you need to do is have a success rate of ~2% to be able to turn a profit.

In the mean time, I think we shouldn't make decisions based on hearsay, when if it was happening it would be quite apparent and demonstrable. 

And I understanding braking this flow makes side-chain/offchain/centralizated-services/etc  more attractive; but I really would greatly appreciated if you could hold off.

Sjors commented:
@petertodd @RHavar this experiment would be interesting to watch, but doesn't tell us much about the actual occurrence.

@RHavar not having RBF by default increases fees for the entire network, yet only benefits 0-conf services. It's not fair to only focus on second layer stuff.

RHavar commented:
>  Being stuck for literal weeks with no recourse(other than "run Peter Todd's full rbf node/tools") is worse in my estimation

That's a bit stretched, if there's change they spend form it to prioritize the transaction. If there's no change, bumpfee doesn't even currently work. 

> but doesn't tell us much about the actual occurrence.

Well if he can profitably take advantage of it, it means it happens more than 2% of the time -- and I will concede that 0-conf is already sufficiently broken that it's not worth preserving.

> not having RBF by default increases fees for the entire network, yet only benefits 0-conf services.

There's certainly a huge amount of potential benefits of a wallet that does full RBF. I think a great example of that would be if your wallet generated ~10 or so transactions, each with a higher fee (and higher nLockn time) according to a schedule so it could automatically bump transactions until they confirm, ensuring you can pay the least possible fees.

Same thing about reliable feebumping, and reverting transactions. 

If these benefits were actually realized, I think it would might be appropriate to make replaceability opt-out. I think right now it's like shooting your horse because in 10 years cars will replace them.

Sjors commented:
> I think a great example of that would be if your wallet generated ~10 or so transactions, each with a higher fee (and higher nLockn time) according to a schedule so it could automatically bump transactions until they confirm, ensuring you can pay the least possible fees.

Yes, this would be very cool. I don't know if it requires keeping your node running for the full duration, but at least the signing could be done in one go. Perhaps a first step woud be an RPC command like "fee_burst_send [amount] [max fee] [deadline]".

instagibbs commented:
There is no currently coded "auto-CPFP" currently, and debate on how to do that as a feature. There are very manual ways of doing this, sure.

The only persuasive argument here is that some users depositing in 0-conf second layer systems will have to wait(and only if they have no change since they can rbf and unmark it). Malicious users can't even double-spend using this without additional tooling.

I think it's a bad tradeoff. but I'll let others weigh in instead.

Sjors commented:
@instagibbs are there second layer proposals where using anything less than 6 confirmations is considered even remotely secure? I think @RHavar reference to second layer solutions referred to a preference for using second layer systems themselves for fast transactions, rather than supporting fast transaction on the layer 1 (using zero conf without opt-in RBF).

instagibbs commented:
@Sjors for better or for worse, many exchanges take <6 confirms, and some services take 0-conf. I do not begrudge this.

I was using "second layer systems" as anything built on top, including Bitcoin casinos.

Sjors commented:
@instagibbs I'll let @RHavar clarify, but I would suggest calling things like Lightning "second layer", while calling exchanges and casinos "off-chain", just for clarity.

RHavar commented:
I use the same definition as @instagibbs which I think is pretty universal. I think him and I agree about everything except which trade-off is worth making.

> are there second layer proposals where using anything less than 6 confirmations is considered even remotely secure?

Sure. Why not? 6 isn't some magical number. It's up to everyone to decide what risk tolerance to accept (which will greatly depend on how miners are behaving). FWIW I've processed 60991 BTC of anonymous deposits requiring 1 conf, and never had any issues. 

If you think you can abuse this, you have my full permission to abuse up to 20 bitcoin in a double-spend against bustadice.com (and consider it a bounty) =)

> There is no currently coded "auto-CPFP" currently, and debate on how to do that as a feature. There are very manual ways of doing this, sure.

Sure, and also the addition of "first seen safe RBF" is something I'd get behind too. 99% of the time this is just as efficient as full RBF  (assuming you don't do off-peak consolidation, and need to eventually consolidate your unspent anyway)

Sjors commented:
>> are there second layer proposals where using anything less than 6 confirmations is considered even remotely secure?

> I use the same definition as @instagibbs which I think is pretty universal. I think him and I agree about everything except which trade-off is worth making.

Using the same term for things like Lightning and custodial services  makes no sense to me. They have completely different threat models. 

> And I understanding braking this flow makes side-chain/offchain/centralizated-services/etc more attractive; but I really would greatly appreciated if you could hold off.

That's the remark I was referring to when I asked my question about 6 confirmations. Custodial services can certainly do 0 conf, but my question is if anything like side-chains or lightning does.

Sjors commented:
@RHavar I'm also not allowed to use your website where I live, due to some effective lobbying by a state owned corporation with a monopoly on gambling. :-)

RHavar commented:
> Using the same term for things like Lightning and custodial services makes no sense to me. 

You're free to use a more specific term

> That's the remark I was referring to when I asked my question about 6 confirmations. Custodial services can certainly do 0 conf, but my question is if non-custodial second layer solutions like side-chains or lightning do.

I'm not sure it's a distinction worth making. There's no universal safe or unsafe. I'd be happy right now to accept a payment from a payment channel where the source has a single confirmation (and even 0, if everything looked sane). Other people might want 6. If a sizable amount of miners were acting adversarially, I wouldn't touch anything with out a hell of a lot more than that.

Being custodial or not doesn't really make much of a difference, AFAICT

petertodd commented:
@RHavar Looks like your site is down.

RHavar commented:
> @RHavar Looks like your site is down.

They're all definitely running. Can you email me ryan@bustabit.com what you're seeing (to take it off this thread)? I don't see any issue in any of my sites. There however is a whole bunch of countries that do block it though, so you might want to try use a VPN if that applies to you. It's also possible there might be an issue with the cloudflare edge you're connecting to, but I doubt it

laanwj commented:
Changing defaults always gives so much discussion, wasted IMO.

Sjors commented:
@laanwj if there's no agreement on this PR, I can point that out in the #11556 discussion. I'll still try to address the remaining code quality issues.

laanwj commented:
Yes, good idea to leave the controversial part in a separate PR at least :)

jonasschnelli commented:
What's the state of this PR?
I think doing it GUI only makes sense for now.

Code-wise, some things are unaddressed:
* Move code away from core classes. I think the only change you want to have in the non GUI files is that help line for `-walletrbf` (we should mention that the default value for QT is 1).
* Try to set `-walletrbf=1` in `src/qt/bitcoin.cpp` (`GuiMain`) or in `createOptionsModel()` in case it's unset.

Sjors commented:
I pushed a new version. This led to a conflict, so I also rebased.

* QT now processes `-walletrbf`, so I no longer need to change the non-GUI files. The exception is `DEFAULT_WALLET_QT_RBF`, which is needed by `wallet/init.cpp` to print the help message. There might be a convoluted way to avoid that, by using `HelpMessageMode=HMM_BITCOIN_QT` to suppress `-walletrbf` and instead show this flag under `UI Options`.

* I'm not sure if `qt/sendcoinsdialog.cpp` is the right place for `gArgs.GetBoolArg("-walletrbf", DEFAULT_WALLET_QT_RBF)`, but that's the same pattern `qt/guiutil.cpp` and `qt/intro.cpp` use.
  * @jonasschnelli did you mean `BitcoinApplication` instead of `GuiMain`? I'm not sure where that would go, and how to then access that from `sendcoinsdialog` (bitcoin.cpp doesn't have a header file).
  * I looked into `OptionsModel`, but as far as I understand what that class is doing, it seems more suitable for global settings that can be toggled. E.g. the display currency can be toggled by the user through a setting, but there's no setting for the user to enable whether or not the RBF checkbox should be on by default. I could make it remember the last choice, but that would increase the scope of this PR. 

* @MarcoFalke I removed `getDefaultWalletRbf` from `WalletModel`

* Added a release note.

Sjors commented:
I’ll fix the linter issue.

TheBlueMatt commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -176,15 +177,16 @@ void SendCoinsDialog::setModel(WalletModel *_model)
 
         // set default rbf checkbox state
-        ui->optInRBF->setCheckState(model->getDefaultWalletRbf() ? Qt::Checked : Qt::Unchecked);
+        const bool fWalletQtRbf = gArgs.GetBoolArg("-walletrbf", DEFAULT_WALLET_QT_RBF);
```
I believe we cache a bunch of the states of other UI elements in the send dialog (eg feerate selectiton), is it worth doing something similar here?

Sjors replied:
Do you mean remembering what the user selected before? Does that use the OptionsModel?

TheBlueMatt replied:
I believe it uses the OptionsModel, yes.

TheBlueMatt reviewed (commented):
Concept ACK changing the Qt default - the usability improvement from providing RBF is very significant, and the only users who may be adversely affected are those paying to services that take 0conf, which appear to be ever fewer (with the exception of in-person shops, which likely will take it either way).

RHavar commented:
> and the only users who may be adversely affected are those paying to services that take 0conf, which appear to be ever fewer 

This is simply not true, and hugely dismissive of the real impacts this change will cause. 

Just today I sent an organization $10k USD by paying them through coinbase. Coinbase guaranteed the exchange rate as long as they saw the transaction on the network with 15 minutes (even if it takes a few hours to confirm). It's a great user experience for all involved (I know exactly how much bitcoin to send) and the seller knows exactly how much USD they will get (after it's all confirmed, and what not) and coinbase gets their cut.

Now if you throw in replaceable transactions, everything falls apart. I can trivially revert the transaction if the exchange rate moves in a way beneficial for me (and even tell the vendor I will resend the funds as a diff transaction). 

I mean, the way things are going eventually bitcoin is going to be a pure settlement layer and none of this will really matter. But it's not nearly that yet, so I just don't get where all this motivation to vandalize bitcoin payments is coming from.

The concern is ostensibly for user-experience, so while I realize time and energy isn't fungible -- but it would be so much better spent on improving other things (like CPFP workflow, and FSS-RBF and things without such negative effects)

TheBlueMatt commented:
@RHavar Sadly, neither CPFP nor FSS-RBF are realistic replacements (CPFP has real additional cost, and FSS-RBF, at least as originally proposed, is highly restrictive on types of transactions that can be replaced). Given the community's seeming unwillingness to actually work together, those targeting very short confirmation windows almost certainly don't have any options *except* RBF right now that have reasonable usability. There may be cases where not using RBF is preferable, and I've suggested many times that people who want that put something like no125=1 in their URIs do so, and we should honor such things, but I dont see a mad rush to do that, plus users can easily still turn it off. But for average payments (the vast majority of which require some confirmations either way), RBF is such a better user experience that having it off by default just causes issues for users.

If something like this gets turned on more commonly, and major providers see an issue, I'm sure they'd be happy to look into URI options to inform clients to turn it off. But, in general, my daily wallet sets RBF off for all payments and all the exchange-rate-guarantee folks seem to have no problem taking my transactions, I assume because the exchange-rate-movement issues just dont appear in the wild (and if they did and they needed something immediately, I'm *sure* there are lots of folks who'd happily offer short-term options contracts to them at scale for a fraction of the fee someone like Coinbase takes in on payments).

petertodd commented:
@RHavar In practice, Coinbase appears to accept opt-in RBF payments just fine for the times I've used them recently (buying hotel/plane tickets via CheapAir, and five minutes ago to buy some Reddit Gold as a test). They of course are just one of many reasons why people send Bitcoin; even if Coinbase has the behavior you mention - which doesn't appear to be true - most Bitcoin exchanges won't even consider a deposit until it has confirmed under any circumstance, making the point irrelevant with them.

Bitcoin Core is used by a relatively technical audience, so I see no harm in some users possibly wanting to untick a checkmark a minority of the time.

Sjors commented:
Rebased after #11556 was merged. 

@TheBlueMatt I created a placeholder issue for your `no125=1` suggestion.

I'll look into using `OptionsModel` to remember the users previous choice. Not sure if that should be blocking though.

Sjors commented:
Travis fails, but `test/functional/zapwallettxes.py` passes for me locally.

If the user selects "Subtract fee from amount", this selection is not remembered for the next transaction. Same goes for selecting e.g. mBTC instead of BTC. It seems that OptionsModel is (mostly) for settings that can be changed through the Options dialog. E.g you can change the default display currency in that dialog and that setting is not affected by selecting a different currency for a specific transaction.

So I don't think we should remember the users previous choice for using RBF.

TheBlueMatt commented:
Kicked travis as it seems almost certainly unrelated, but #11831 should fix the race, I believe.

TheBlueMatt commented on src/qt/forms/sendcoinsdialog.ui:
```diff
@@ -1106,12 +1106,12 @@
              </property>
              <property name="toolTip">
-              <string>This allows you to increase the fee later if the transaction takes a long time to confirm. This will also cause the recommended fee to be lower. ("Replace-By-Fee", BIP 125)</string>
+              <string>Not using Replace-By-Fee (BIP-125) marks the transaction as final. This will also cause the recommended fee to be higher.</string>
```
Heh, I'm less a fan of the copy here now. Maybe "Not using Replace-By-Fee (BIP-125) prevents you from increasing a transaction's fee after it is sent. This may cause the recommended fee to be higher to compensate for the increased risk a transaction is delayed." or something like that?

Sjors replied:
I changed it to:
> Without Replace-By-Fee (BIP-125) you can't increase a transaction's fee after it is sent. A higher fee may be recommended to compensate for increased transaction delay risk.

morcos commented:
Sorry for chiming in late, but a couple of comments:

1) Concept ACK. I'm strongly in favor of making RBF default to on, I realize there are some downsides, but I think they are significantly outweighed by the fact that accurate fee estimation is essentially an impossible task and being able to adjust fee afterwards is the single biggest usability improvement we can offer now.

2) I think we should flip RPC and QT at the same time.  I recognize there are some small risks with flipping the RPC default, but that can just be release noted.  Separating them is just introducing unnecessary configuration complexity.  It will also be confusing to users that use RPC while running QT and have their transactions go out differently.  Fundamentally the easiest and best thing to do is just change DEFAULT_WALLET_RBF  to true.

3) I question flipping the meaning of the checkbox, I realize it's kind of silly for it to be always defaulted to checked, but are we not worried some people will just be used to expecting the checkbox to be checked for RBF?  I'm fine being outvoted on this aspect though if we think its short term pain for long term gain.

Sjors commented:
@RHavar can you add your five cents to the [bitcoin-dev list discussion](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2017-December/015352.html) about a `-no150` flag? In particular regarding @luke-jr's point.

@MarcoFalke:

2 - The additional complexity is only three lines of code. It would seem trivial to switch the RPC default over one version later, when people are more familiar with it. The 0.16 release notes (assuming this makes it in) could warn that that's coming in the next version.

Regarding:
> confusing to users that use RPC while running QT and have their transactions go out differently": 

That shouldn't be the case. 

3 - I'm fine with changing the word "disable" to "enable" in the checkbox and checking it by default. That wouldn't require changing any other copy, not even the tooltip. Any objections?

jonasschnelli commented:
I kinda have to agree with @morcos. My first review also mentioned that we shouldn't to a distinction between RPC/GUI (https://github.com/bitcoin/bitcoin/pull/11605#issuecomment-342010545).

I'm fine with both, but would prefer a complete default switchover.

RHavar commented:
With a heavy heart, I'm going to drop my objection to this change. Bitcoin as a payment network is pretty much dead (congratulations boys!) so there is becoming little point in trying to protect 0-confs. Low-balling fees and fee-bumping will lead to a healthier fee market, so might as well do it.

It should also be imho considered a very high priority to fix `bumpfee` to handle the case without change. (Contrary to popular belief this isn't the edge case, but rather the norm (~80%) when simulating the transaction load a bitcoin service gets when using  non-naive (read: proper constraint optimizer) input (coin) and output (which of the batch to send) selection ).

Sjors commented:
@RHavar thanks. Among my ever growing todo list is to make the bump fee functionality better, especially in the UI. There was some [recent IRC discussion](https://botbot.me/freenode/bitcoin-core-dev/2017-12-11/?msg=94528710&page=4) about this as well.

I'm still reluctant about changing the RPC default, which we also [chatted about](). Once thing I'd like to (see someone else) do is go through popular wallets and services to see how they treat RBF payments and how they treat bumped fees. When it's a relatively small user base, Bitcoin Core QT users, that's one thing. At least the sender probably knows a bit more than average about Bitcoin and can community directly with the recipient. But when every Bitcoin ATM's or other automated process that runs Core starts doing this there's a problem.

laanwj commented on src/wallet/init.cpp:
```diff
@@ -31,7 +31,7 @@ std::string GetWalletHelpString(bool showDebug)
     strUsage += HelpMessageOpt("-spendzeroconfchange", strprintf(_("Spend unconfirmed change when sending transactions (default: %u)"), DEFAULT_SPEND_ZEROCONF_CHANGE));
     strUsage += HelpMessageOpt("-txconfirmtarget=<n>", strprintf(_("If paytxfee is not set, include enough fee so transactions begin confirmation on average within n blocks (default: %u)"), DEFAULT_TX_CONFIRM_TARGET));
-    strUsage += HelpMessageOpt("-walletrbf", strprintf(_("Send transactions with full-RBF opt-in enabled (default: %u)"), DEFAULT_WALLET_RBF));
+    strUsage += HelpMessageOpt("-walletrbf", strprintf(_("Send transactions with full-RBF opt-in enabled (default: %u for RPC, %u for QT)"), DEFAULT_WALLET_RBF, DEFAULT_WALLET_QT_RBF));
```
Please use "GUI" not "QT" when referring to the GUI in messages.

Sjors replied:
Will fix.

laanwj commented:
I'm starting to warm up to the idea of changing this default. Seems it would make the fees escalate a little less if clients use more conservative estimates.

Sjors commented:
I flipped the checkbox: it's now checked and says "Enabled" and adjusted the tooltip. This passes the strobe test better (removes a bunch of double negatives). See new screenshots.

I renamed the variable `enableRBF` to `enableRBF`.

@laanwj suggested:
> What about a separate option for the GUI, GUI-only (so documented in utilitydialog.cpp instead of init.cpp), that defaults to -walletrbf? This would allow it to be completely disentangled from the core.

E.g. `-guiwalletrbf`? I'm worried that would just make it more confusing, unless we also rename `-walletrbf` to `-rpcwalletrbf`.

I'm generally in favor of treating the RPC and GUI as different products, since they target very different users. The only factor that should constrain that is code complexity. From that perspective it would make sense to have different defaults and use `-walletX` vs. `-rpcX` settings.

I can't think of many  examples outside RBF where this would make sense. Only memory usage comes to mind; a GUI could use a lot more than a daemon, since people either don't leave it open all the time, or at least are used to quitting applications when their computer gets too slow.

It also seems quite easy to add `-guiwalletrbf` at a later stage and undo the entanglement.

laanwj commented:
> E.g. -guiwalletrbf? I'm worried that would just make it more confusing, unless we also rename -walletrbf to -rpcwalletrbf.

The difference is that here, it's effectively two different things already. One changes the default in the GUI, one changes the default that's used with RPC. There is no overlap in code between what the options do.

I really dislike the `(default: %u for RPC, %u for GUI)` help message because it leaves one thing open: what if you use the GUI, but type commands in the debug console?

And I agree with regard to `-rpcwalletrbf` but it's too late for that.

At least the GUI option could be a nice setting in the options dialog. GUI users tend to prefer that to adding things to bitcoin.conf.

Sjors commented:
Perhaps we can add `-rpcwalletrbf` and alias for `-walletrbf` and deprecate the latter (with zero rush).  I could then replace `(default: %u for RPC, %u for GUI)` with a warning that this setting applies to RPC only. And then add `-guiwalletrbf`.

Sjors commented:
I implemented `-guiwalletrbf`, added `-rpcwalletrbf` and a deprecation warning for `-walletrbf`. I added [RPC] to the commit name. This issue probably needs an RPC tag now.

RPC and debug console won't use RBF by default and honor the `-rpcwalletrbf` (or `-walletrb`) flag.

laanwj commented:
> RPC and debug console won't use RBF by default and honor the -rpcwalletrbf (or -walletrb) flag.

Sounds good to me, as it's documented. Doing it otherwise would introduce new complications between core and GUI code that aren't worth it.

I'm not sure the deprecation is worth it either. Soon enough I expect that RBF will be the default for both GUI and RPC, and there's no longer need for two separate options nor defaults.

Sjors commented:
I pushed some whitespace and variable name fixes, as well as repaired the `bumpfee.py` test.

I'm open to not doing the rename / deprecation, but if there's not too much downside to it, I do think it adds clarity.

jonasschnelli commented:
I think the consensus is that we enabled RBF by default now. Merging this as it is would cause confusion. I think we should move the PR towards enabling it everywhere.

Sjors commented:
@jonasschnelli I think this needs to be addressed first (copied from above):

> I'm still reluctant about changing the RPC default, which we also [chatted about](https://botbot.me/freenode/bitcoin-core-dev/2017-12-11/?msg=94528264&page=4). Once thing I'd like to (see someone else) do is go through popular wallets and services to see how they treat RBF payments and how they treat bumped fees. When it's a relatively small user base, Bitcoin Core QT users, that's one thing. At least the sender probably knows a bit more than average about Bitcoin and can community directly with the recipient. But when every Bitcoin ATM's or other automated process that runs Core starts doing this there's a problem.

jonasschnelli commented:
Merging this would mean we have to get rid of the -rpcwalletrbg -guiwalletrbf split later?

Sjors commented:
No, we would get rid of the `-*walletrbf` flag entirely. If there's any reason to keep that flag, then it needs to be separate for QT and GUI.

jonasschnelli commented:
I don't think a split between GUI and RPC defaults is desirable for something we want to have enabled by default long term anyways (on both sides). Splits between GUI and RPC should be avoided where possible to lower confusion.

Sjors commented:
If we can't separate the behavior, I don't think we should change the default at all, until we understand what the impact is on automated services with large transaction volumes. Or unless we're sure they read the release notes.

jonasschnelli commented:
I think if you run large volume business, you either use Qt or much more likely headless/RPC. You certainly will (should!) read default changes in the release notes during a software upgrade.

The split use case would be for someone running both "worlds" (GUI/RPC) with different use cases which I think is unlikely.

RHavar commented:
> Seems it would make the fees escalate a little less if clients use more conservative estimates.

Yeah, agree. The current situation is absolutely insane (and tragically foreseeable). While obviously not the cause of the clusterfuck we're in, having clients aggressively blindly outbid each other is not helping.

However, I'm still rather skeptical this is anything more than hack. If you want people to low ball fees, and then progressively bump them -- it really should be reliable. In my wallet now I have dozens of transactions that core isn't capable of fee-bumping due to either no change, or the receiver has spent/swept the output.

It would also make sense to build this more first class. e.g. Generate 10 transactions with a progressively increasing nlockTime as well as fee rate. (although this is starting to get a bit off topic)

> If we can't separate the behavior, I don't think we should change the default at all, until we understand what the impact is on automated services with large transaction volumes

There's not going to really be any. A year ago or so, it would've been very different when a number of block-explorers and stuff used to reject them (until they confirmed) or show scary warnings. But there's nothing like that today.  (Although they still show scary warnings if you actually do a fee bump, warning about double-spends)

Sjors commented:
Alright, so we had some more back and forth on IRC. The new version I just pushed:
* drops `-guiwalletrbf`; the GUI will just use RBF by default
* keeps the original `-walletrbf` flag, off by default, but now now only impacts RPC

jonasschnelli commented:
Nice.
utACK dcd50b90b94fbf2295a122693570218932bf15b4

Consider also updating the original PR description. It will be part of the git history once it's merged.

Sjors commented:
@jonasschnelli fixed the PR description.

laanwj commented:
utACK dcd50b90b94fbf2295a122693570218932bf15b4

laanwj commented:
Travis fails on trailing whitespace :(
```
diff --git a/doc/release-notes.md b/doc/release-notes.md
@@ -69,0 +70,8 @@ will only create hierarchical deterministic (HD) wallets.
+The send screen now uses BIP-125 RBF by default, regardless of `-walletrbf`.
+There is a checkbox to mark the transaction as final.
^---- failure generated from contrib/devtools/lint-whitespace.sh
```

Sjors committed:
[Wallet] Use RBF by default in QT only

GUI wallet uses RBF by default, regardless of -walletrbf.

RPC and debug console in the GUI remain unchanged; they don't
use RBF by default, unless launched with -walletrbf=1.

Sjors commented:
@laanwj fixed

Atom either automatically fixes all whitespace in a file or nothing at all. I haven't been able to tweak it to only fix whitespace in lines that I'm working on, as per the code style guidelines.

Running `TRAVIS_COMMIT_RANGE=91eeaa...5cbbbd71 contrib/devtools/lint-whitespace.sh` locally is also not super practical, though could probably be automated.

Installing linters helps spotting missing whitespace in time, so that mitigates the issue (I hadn't done that yet on my new machine). Although in the case of [linter-markdown](https://atom.io/packages/linter-markdown), it takes a few additional steps to make it care about whitespace. I might make a PR with some hints later.

Sjors commented:
Travis still unhappy, not sure why:
<img width="892" alt="schermafbeelding 2017-12-22 om 10 04 53" src="https://user-images.githubusercontent.com/10217/34292078-9561546c-e6ff-11e7-8a47-84f362251c63.png">

laanwj commented:
Seems Travis is happy now, probably a transient issue.

laanwj commented:
Tested ACK 5cbbbd7.
